apps/studio/src/components/ResourceSelector/ResourceSelector.tsx

```diff
@@ -1,157 +1,228 @@
-import { Suspense, useState } from "react"
-import {
-  Box,
-  Flex,
-  HStack,
-  Icon,
-  Skeleton,
-  Spacer,
-  Text,
-} from "@chakra-ui/react"
+import { Suspense, useEffect, useState } from "react"
+import { Box, Flex, HStack, Skeleton, Spacer, Text } from "@chakra-ui/react"
 import { Button, Link } from "@opengovsg/design-system-react"
-import { ResourceType } from "~prisma/generated/generatedEnums"
+import { ResourceType } from "@prisma/client"
 import { BiHomeAlt, BiLeftArrowAlt } from "react-icons/bi"
 
+import type { PendingMoveResource } from "~/features/editing-experience/types"
+import type { ResourceChildrenOfType } from "~/schemas/resource"
+import { useQueryParse } from "~/hooks/useQueryParse"
+import { sitePageSchema } from "~/pages/sites/[siteId]"
 import { trpc } from "~/utils/trpc"
 import { ResourceItem } from "./ResourceItem"
 
+interface ResourceSelectorProps {
+  onChange: (resourceId: string) => void
+  selectedResourceId?: string
+  existingResource?: PendingMoveResource | null
+  onlyShowFolders?: boolean
+}
+
+const generatePermalink = (parents: PendingMoveResource[]) => {
+  return parents.map((parent) => parent.permalink).join("/")
+}
+
 const SuspensableResourceSelector = ({
-  siteId,
-  selectedResourceId,
   onChange,
-  isDisabledFn,
+  selectedResourceId,
+  existingResource,
+  onlyShowFolders = false,
 }: ResourceSelectorProps) => {
```

Comment:
thought: I think I relooked at this piece of code a couple of times for this review and it is quite a bit of things that is going on within this single component. Putting this as a thought because it is non-blocking in nature, but would be good to refactor this component (either in this PR directly or in a subsequent one).

Main points:
1. The `resourceStack` state is the data structure that is holding most of the information needed to get the selector working as intended, but we are currently manipulating this state directly, rather than exposing it through more understandable APIs like sink and lift (or push and pop, etc).
    1. Recommendation: To use a custom hook that manages this state and exposes only the APIs from the hook. From what I see, the APIs exposed should be the sink and lift calls, moveDest, parentDest and curResourceId.
2. The second `useEffect` to call the `onChange` still feels weird to me, as it still seems to be synchronising states.
    1. Recommendation: We can probably pass the `onChange` function to the custom hook suggested in 1, then let the custom hook handle calling this `onChange` whenever the resource stack changes, so we don't need to rely on the `useEffect` to synchronise the states.

Reply:
yup actually this sounds good!

Reply:
attempted a refactor into `useResourceStack` custom hooks!